sdk/PrebidMobile/PrebidURLProtocol.m

```diff
@@ -0,0 +1,92 @@
+/*   Copyright 2017 APPNEXUS INC
+ 
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+ 
+ http://www.apache.org/licenses/LICENSE-2.0
+ 
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import "EGOCache.h"
+#import "PrebidURLProtocol.h"
+
+@interface PrebidURLProtocol () <NSURLConnectionDelegate>
+
+@property (nonatomic, strong) NSURLConnection *connection;
+
+@end
+
+@implementation PrebidURLProtocol
+
++ (BOOL)canInitWithRequest:(NSURLRequest *)request {
+    static NSUInteger requestCount = 0;
+    if ([NSURLProtocol propertyForKey:@"PrebidURLProtocolHandledKey" inRequest:request]) {
+        return NO;
+    }
+    if ([request.URL.absoluteString containsString:@"prebid.adnxs.com/pbc/v1/cache"]) {
+        return YES;
+    }
+    return NO;
+}
+
++ (NSURLRequest *)canonicalRequestForRequest:(NSURLRequest *)request {
+    return request;
+}
+
++ (BOOL)requestIsCacheEquivalent:(NSURLRequest *)a toRequest:(NSURLRequest *)b {
+    return [super requestIsCacheEquivalent:a toRequest:b];
+}
+
+- (void)startLoading {
+    NSMutableURLRequest *newRequest = [self.request mutableCopy];
+    [NSURLProtocol setProperty:@YES forKey:@"PrebidURLProtocolHandledKey" inRequest:newRequest];
+
+    NSRange range = [newRequest.URL.absoluteString rangeOfString:@"uuid="];
+    if (range.location == NSNotFound) {
+        NSLog(@"No uuid in URL");
+    }
+    NSString *uuid = [newRequest.URL.absoluteString substringFromIndex:(range.location + 5)];
+
+    NSDictionary *dict = (NSDictionary *)[[EGOCache globalCache] objectForKey:uuid];
+    if (dict != nil) {
+        NSData *data = [NSJSONSerialization dataWithJSONObject:dict
+                                                       options:NSUTF8StringEncoding
+                                                         error:nil];
+        NSURLResponse* response = [[NSHTTPURLResponse alloc] initWithURL:[self.request URL] statusCode:200 HTTPVersion:@"HTTP/1.1" headerFields:@{@"Access-Control-Allow-Origin": @"https://pubads.g.doubleclick.net", @"Access-Control-Allow-Credentials" : @"true"}];
```

Comment:
Should MoPub origin be added here too?

Reply:
Yup will add. Need to test MoPub still.

Reply:
Added MoPub support with Prebid Cache since we can't use local cache solution as MoPub uses WKWebView.